### PR TITLE
feat: start mgo with arg to use CA cert

### PR DESCRIPTION
**Pull request by wallyworld**
Mongo 4.4.30 requires client connections be properly authenticated.
Some enablement work is need to make juju compatibile with this mongo release.
The mongo service needs to be started with a CA cert to use and client connections need to be configured with certificates.

We'll use the controller's CA cert when starting mongo.
For client connections, issue a new certificate off the CA cert. The client cert is valid for 15 minutes. Once the connection is created, it stays open. If a new connection is needed and the cert has expired, a new one is generated.

Also update the mongo args to use the new "tls" terminology, but only where the mongo version is 4.4. Juju 2.8 uses an older deb mongo version and early 2.9 uses 4.0 and the newer "tls" args were introduced in mongodb 4.1. So we need to stick with the old "ssl" args for those mongo versions.

For pre merge tests, the seed-repository target needed to optionally push the old juju-db 4.4 image so the older juju could bootstrap.

Drive by: remove obsolete client pre-merge tests.

## QA steps

```
juju bootstrap lxd test
juju switch controller
juju enable-ha
juju deploy ubuntu
```
You can exec into the controller and `snap info juju-db` shows `4.4.30` is installed.

Check upgrades from a juju version using mongodb `4.4.24`
```
juju bootstrap lxd --agent-version 2.9.53
juju switch controller
juju upgrade-controller --build-agent
```

The juju-db snap config in `/var/snap/juju-db/common/juju-=db.config` gets regenerated with the new auth related config which is compatible with mongodb `4.4.24`
```
cat juju-db.config 
# WARNING
# autogenerated by juju on 2026-01-15 23:55:41.193281297 +0000 UTC m=+0.906492080
# manual changes to this file are likely to be overwritten
port = 37017
tlsCertificateKeyFile = /var/snap/juju-db/common/server.pem
tlsCertificateKeyFilePassword=ignored = true
slowms = 1000
auth = true
oplogSize = 1024
storageEngine = wiredTiger
quiet = true
keyFile = /var/snap/juju-db/common/shared-secret
journal = true
replSet = juju
dbpath = /var/snap/juju-db/common/db
logpath = /var/snap/juju-db/common/logs/mongodb.log
bind_ip_all = true
tlsMode = requireTLS
ipv6 = true
tlsCAFile = /var/snap/juju-db/common/ca.crt
tlsAllowInvalidHostnames = true
```

On k8s

```
make microk8s-operator-update
juju bootstrap microk8s test
juju switch controller
juju deploy snappass-test
```

On older mongo
Get a juju 2.8 cli client
```
./juju bootstrap lxd
juju switch controller
juju ssh 0
cat /etc/systemd/system/juju-db.service
...
ExecStart=/usr/bin/mongod --auth --bind_ip_all --dbpath '/var/lib/juju/db' --ipv6 --journal --keyFile '/var/lib/juju/shared-secret' --oplogSize 1024 --port 37017 --quiet --replSet juju --slowms 1000 --sslMode requireSSL --sslPEMKeyFile '/var/lib/juju/server.pem' --sslPEMKeyPassword=ignored --storageEngine wiredTiger --syslog

juju upgrade-controller --build-agent
juju ssh 0
cat /etc/systemd/system/juju-db.service
...
ExecStart=/usr/bin/mongod --auth --bind_ip_all --dbpath '/var/lib/juju/db' --ipv6 --journal --keyFile '/var/lib/juju/shared-secret' --oplogSize 1024 --port 37017 --quiet --replSet juju --slowms 1000 --sslAllowInvalidHostnames --sslCAFile '/var/lib/juju/ca.crt' --sslMode requireSSL --sslPEMKeyFile '/var/lib/juju/server.pem' --sslPEMKeyPassword=ignored --storageEngine wiredTiger --syslog

...

```

## Links

**Jira card:** [JUJU-9072](https://warthogs.atlassian.net/browse/JUJU-9072)

[JUJU-9072]: https://warthogs.atlassian.net/browse/JUJU-9072?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by wallyworld** [comment]
> Code looks good, though I can't replicated the PR steps because when I attempt to bootstrap I get:
> 
> ```shell
> root@juju-92f03c-0:~# systemctl status snap.juju-db.*
> ● snap.juju-db.logrotate.timer - Timer logrotate for snap application juju-db.logrotate
>      Loaded: loaded (/etc/systemd/system/snap.juju-db.logrotate.timer; disabled; vendor preset: enabled)
>      Active: active (waiting) since Mon 2026-01-12 09:17:41 UTC; 3min 58s ago
>     Trigger: Tue 2026-01-13 00:00:00 UTC; 14h left
>    Triggers: ● snap.juju-db.logrotate.service
> 
> Jan 12 09:17:41 juju-92f03c-0 systemd[1]: Started Timer logrotate for snap application juju-db.logrotate.
> 
> ● snap.juju-db.daemon.service - Service for snap application juju-db.daemon
>      Loaded: loaded (/etc/systemd/system/snap.juju-db.daemon.service; enabled; vendor preset: enabled)
>     Drop-In: /etc/systemd/system/snap.juju-db.daemon.service.d
>              └─overrides.conf
>      Active: failed (Result: exit-code) since Mon 2026-01-12 09:17:43 UTC; 3min 56s ago
>     Process: 5060 ExecStart=/usr/bin/snap run juju-db.daemon (code=exited, status=14)
>    Main PID: 5060 (code=exited, status=14)
>         CPU: 175ms
> 
> Jan 12 09:17:43 juju-92f03c-0 systemd[1]: snap.juju-db.daemon.service: Scheduled restart job, restart counter is at 4.
> Jan 12 09:17:43 juju-92f03c-0 systemd[1]: Stopped Service for snap application juju-db.daemon.
> Jan 12 09:17:43 juju-92f03c-0 systemd[1]: snap.juju-db.daemon.service: Start request repeated too quickly.
> Jan 12 09:17:43 juju-92f03c-0 systemd[1]: snap.juju-db.daemon.service: Failed with result 'exit-code'.
> Jan 12 09:17:43 juju-92f03c-0 systemd[1]: Failed to start Service for snap application juju-db.daemon.
> ```
> 
> Here is the output from the bootstrap:
> 
> ```
> Installing Juju machine agent
> 2026-01-12 09:17:35 INFO juju.cmd supercommand.go:56 running jujud [2.9.54.1 2882471b366bfaabdd1f26fb04c5f2ecf7b58321 gc go1.25.5]
> 2026-01-12 09:17:35 INFO juju.agent identity.go:22 writing system identity file
> 2026-01-12 09:17:35 ERROR juju.mongo mongo.go:654 could not set the value of "/sys/kernel/mm/transparent_hugepage/defrag" to "never" because of: open /sys/kernel/mm/transparent_hugepage/defrag: permission denied
> 2026-01-12 09:17:35 ERROR juju.mongo mongo.go:654 could not set the value of "/proc/sys/net/core/netdev_max_backlog" to "1000" because of: "/proc/sys/net/core/netdev_max_backlog" does not exist, will not set "1000"
> 2026-01-12 09:17:35 ERROR juju.mongo mongo.go:654 could not set the value of "/sys/kernel/mm/transparent_hugepage/enabled" to "never" because of: open /sys/kernel/mm/transparent_hugepage/enabled: permission denied
> 2026-01-12 09:17:35 WARNING juju.mongo mongo.go:484 overwriting args.dataDir (set to /var/lib/juju) to /var/snap/juju-db/common
> 2026-01-12 09:17:35 INFO juju.mongo mongo.go:489 Ensuring mongo server is running; data directory /var/snap/juju-db/common; port 37017
> 2026-01-12 09:17:35 WARNING juju.mongo service.go:338 configuring mongod  with --noauth flag enabled
> 2026-01-12 09:17:35 INFO juju.packaging manager.go:103 installing "juju-db" via "snap"
> 2026-01-12 09:17:35 INFO juju.packaging.manager run.go:88 Running: snap install  --channel 4.4/candidate juju-db
> 2026-01-12 09:22:47 INFO juju.worker.peergrouper initiate.go:68 finished InitiateMongoServer
> ERROR failed to start mongo: cannot initiate replica set: cannot dial mongo to initiate replicaset: no reachable servers
> ERROR failed to bootstrap model: subprocess encountered error code 1
> ```

I just tested again with no issue.
Can you look at the content of `/var/snap/juju-db/common/logs/mongodb.log`?
Also, the content of `/var/snap/juju-db/common/juju-db.config` should be
```
# WARNING
# autogenerated by juju on 2026-01-12 23:06:20.51554226 +0000 UTC m=+1.695387613
# manual changes to this file are likely to be overwritten
bind_ip_all = true
auth = true
tlsAllowInvalidHostnames = true
oplogSize = 1024
tlsMode = requireTLS
tlsCertificateKeyFile = /var/snap/juju-db/common/server.pem
tlsCertificateKeyFilePassword=ignored = true
port = 37017
tlsCAFile = /var/snap/juju-db/common/ca.crt
storageEngine = wiredTiger
replSet = juju
dbpath = /var/snap/juju-db/common/db
logpath = /var/snap/juju-db/common/logs/mongodb.log
keyFile = /var/snap/juju-db/common/shared-secret
journal = true
slowms = 1000
quiet = true
ipv6 = truer
```

File: mongo/open.go [hunk 1]
```diff
@@ -129,12 +135,33 @@ func DialInfo(info Info, opts DialOpts) (*mgo.DialInfo, error) {
...
+			return nil, errors.Annotate(err, "making juju-db client certificate")
+		}
+		cert, key, err := leaf.ToPemParts()
+		if err != nil {
+			return nil, errors.Annotate(err, "encoding juju-db client certificate to pem")
+		}
+		clientCert, err := tls.X509KeyPair(cert, key)
+		if err != nil {
+			return nil, errors.Annotate(err, "parsing juju-db client certificate")
+		}
```
Comment (hpidcock) [inline mongo/open.go]: This is really expensive, generating a new RSA3072 keypair, plus the signing of the certificate. Lastly, the certificate generated has a 10 year validity period.

Since `DialInfo` and `DialWithInfo` are called from the state worker when it starts, if it is bouncing it would issue a lot of certificates. Are we able to generate this certificate when the agent starts up instead? It would also help for tests that rely on this, so they can also pass in a pre-generated certificate.

Comment (wallyworld) [reply mongo/open.go]: RE: validity period - that's what juju has used since day 1 for all the certs it creates (which no one likes). We'd need to agree on any change.

I can move the code - it's only called once on startup to initialise the state pool but as you say, we'd do it again each time if the worker bounced.

Comment (wallyworld) [reply mongo/open.go]: client cert is now generated (once) in the machine agent

Comment (wallyworld) [reply mongo/open.go]: and the client certs get a NotAfter time giving them15 minutes validity

**Comment by sinanawad** [comment]
@SimonRichardson Please take another look, you and @hpidcock are reviewing this.

File: cloudconfig/podcfg/image.go [hunk 2]
```diff
@@ -29,14 +29,15 @@ func (cfg *ControllerPodConfig) GetControllerImagePath() (string, error) {
 
 func (cfg *ControllerPodConfig) mongoVersion() (*mongo.Version, error) {
 	snapChannel := cfg.Controller.JujuDBSnapChannel()
-	vers := strings.Split(snapChannel, "/")[0] + ".0"
+	vers := strings.Split(snapChannel, "/")[0]
```
Comment (SimonRichardson) [inline cloudconfig/podcfg/image.go:32]: This will fail if for any reason the snap channel is set to "4.4". You'll get the error:

```sh
invalid version "4.4"
```

Even though 4.4 is considered bad, it's still in the wild, so we might want to consider still supporting this version string.

File: cloudconfig/podcfg/podcfg_test.go [hunk 3]
```diff
@@ -71,13 +71,13 @@ func (*podcfgSuite) TestOperatorImagesDefaultRepo(c *gc.C) {
 	c.Assert(path, gc.Equals, "jujusolutions/jujud-operator:6.6.6.666")
 	path, err = podConfig.GetJujuDbOCIImagePath()
 	c.Assert(err, jc.ErrorIsNil)
-	c.Assert(path, gc.Equals, "jujusolutions/juju-db:4.4")
+	c.Assert(path, gc.Equals, "jujusolutions/juju-db:4.4.30")
 }
```
Comment (SimonRichardson) [inline cloudconfig/podcfg/podcfg_test.go:75]: Keeping a 4.4 test would have show this error.

File: cmd/jujud/agent/machine.go [hunk 4]
```diff
@@ -941,9 +947,34 @@ func mongoDialOptions(
...
 	dialOpts.PostDialServer = mongoDialCollector.PostDialServer
+	dialOpts.GenerateClientCertificate = generateClientCert
 	return dialOpts, nil
 }
 
+func (a *MachineAgent) generateClientCert(caCert, caPrivateKey string) (*tls.Certificate, error) {
+	a.mongoCertMutex.Lock()
+	defer a.mongoCertMutex.Unlock()
+
+	if a.mongoClientCert != nil {
```
Comment (SimonRichardson) [inline cmd/jujud/agent/machine.go]:
```suggestion
	if a.mongoClientCert != nil && len(a.mongoClientCert.Certificate) > 0 {
```

Better safe than sorry.

File: cmd/jujud/agent/machine.go [hunk 5]
```diff
@@ -941,9 +947,34 @@ func mongoDialOptions(
...
+		// Check whether cert is close to, or has, expired.
+		c, err := x509.ParseCertificate(a.mongoClientCert.Certificate[0])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if time.Now().Before(c.NotAfter.Add(-1 * time.Minute)) {
```
Comment (SimonRichardson) [inline cmd/jujud/agent/machine.go]: nit: this should be a clock.

And the `-1 * time.Minute` should be a const.

File: cmd/jujud/agent/machine.go [hunk 6]
```diff
@@ -397,6 +399,8 @@ type MachineAgent struct {
 
 	mongoInitMutex   sync.Mutex
 	mongoInitialized bool
+	mongoCertMutex   sync.Mutex
```
Comment (SimonRichardson) [inline cmd/jujud/agent/machine.go]: This mutex could be generalized with the mongoInitMutex, as neither are overlapping. It's not critical though.

File: mongo/service.go [hunk 7]
```diff
@@ -491,6 +539,8 @@ func generateConfig(mongoPath string, oplogSizeMB int, version Version, usingMon
 
 	if usingMongo2 {
 		mongoArgs.SSLOnNormalPorts = true
+	} else if usingMongo44 {
+		mongoArgs.SSLMode = "requireTLS"
```
Comment (SimonRichardson) [inline mongo/service.go:534]: Does this mean that 4.4 that we've currently got as a snap (not 4.4.30 snap) work with TLS already? Because I would have thought it would be safer to say, `usingMongo44_30`?

File: make_functions.sh [hunk 8]
```diff
@@ -154,11 +154,22 @@ build_push_operator_image() {
...
 seed_repository() {
   set -x
-  "$DOCKER_BIN" pull "ghcr.io/juju/juju-db:${JUJU_DB_VERSION}"
-	"$DOCKER_BIN" tag "ghcr.io/juju/juju-db:${JUJU_DB_VERSION}" "${OCI_REGISTRY_USERNAME}/juju-db:${JUJU_DB_VERSION}"
-	"$DOCKER_BIN" push "${OCI_REGISTRY_USERNAME}/juju-db:${JUJU_DB_VERSION}"
+  # seed the juju-db image, maybe including the legacy 4.4 version.
+  seed_repository_jujudb "${JUJU_DB_VERSION}"
+  if [[ -n "${LEGACY_JUJU_DB_VERSION:-}" ]]; then
+    seed_repository_jujudb "${LEGACY_JUJU_DB_VERSION}"
+  fi
```
Comment (SimonRichardson) [inline make_functions.sh]:
```suggestion
  if [[ -n "${LEGACY_JUJU_DB_VERSION:-}" ]]; then
    seed_repository_jujudb "${LEGACY_JUJU_DB_VERSION}"
  else
    seed_repository_jujudb "${JUJU_DB_VERSION}"
  fi
```

Why seed both? How do you know it's actually using the right one? Pick one.

File: mongo/service.go (hunk 7 above)
Comment (wallyworld) [reply mongo/service.go:534]: `TLSxxx` naming (replacing 'SSLxxx` was added in 4.1.
So our existing 4.4.24 snaps from the 4.4 track work just fine.
It's only the very old 4.0 used in juju 2.8 and early 2.9 what needs the ssl args.

File: cloudconfig/podcfg/image.go (hunk 2 above)
Comment (wallyworld) [reply cloudconfig/podcfg/image.go:32]: Sure, added a fix. Shouldn't be needed because upgrades Just Work. But users could explicitly point juju to the incorrect track via config.

File: cmd/jujud/agent/machine.go (hunk 5 above)
Comment (wallyworld) [reply cmd/jujud/agent/machine.go]: I considered it but the MachienAgent struct doesn't have a clock and there's no test for this corner case so it didn't seem worth wiring it all up and then not getting any tangible benefit.

File: make_functions.sh (hunk 8 above)
Comment (wallyworld) [reply make_functions.sh]: The non-legacy one is needed for upgrade

File: mongo/service.go [hunk 9]
```diff
@@ -298,26 +310,60 @@ func (mongoArgs *ConfigArgs) asMap() configArgsConverter {
...
-		pemPassword := mongoArgs.PEMKeyPassword
-		if pemPassword == "" {
-			pemPassword = "ignored"
+		if usingTLSArgs {
+			result["tlsCertificateKeyFile"] = utils.ShQuote(mongoArgs.PEMKeyFile)
+			// --tlsCertificateKeyFilePassword must be concatenated to the equals sign (lp:1581284)
+			pemPassword := mongoArgs.PEMKeyPassword
+			if pemPassword == "" {
+				pemPassword = "ignored"
+			}
```
Comment (manadart) [inline mongo/service.go]: Can move this outside the block.

File: mongo/service.go [hunk 10]
```diff
@@ -298,26 +310,60 @@ func (mongoArgs *ConfigArgs) asMap() configArgsConverter {
...
 		}
-		result["sslPEMKeyPassword="+pemPassword] = flagMarker
 	}
 
 	if mongoArgs.AuthKeyFile != "" {
 		result["auth"] = flagMarker
+		// Juju doesn't create or update certificates with
+		// the SANs set to the replica IP addresses/hostnames
+		// so we need to disable hostname verification.
+		if usingTLSArgs {
```
Comment (manadart) [inline mongo/service.go]: Couldn't we do this check once early to choose a set of keys? Would tidy things up a bit.

Comment (wallyworld) [reply mongo/service.go]: Added a key mapping func.
